doc/source/user-profiling.rst

```diff
@@ -0,0 +1,486 @@
+Profiling for Ray Users
+=======================
+
+This document is intended for users of Ray who want to know how to evaluate 
+the performance of their code while running on Ray. Profiling the 
+performance of your code can be very helpful to determine performance 
+bottlenecks or where your code may not be parallelizing properly. If you 
+are interested in pinpointing why your Ray application may not be 
+achieving speedups as expected, then do read on!
+
+
+A Basic Example to Profile
+--------------------------
+
+Let's try to profile a simple example, and compare how different looping 
+call structures of the same remote function affects performance.
+
+As a stand-in for a computationally intensive and possibly slower function,
+let's define our remote function to just sleep for 0.5 seconds:
+
+.. code-block:: python
+
+  import ray
+  import time
+
+  # Our time-consuming remote function
+  @ray.remote
+  def func():
+    time.sleep(0.5)
+
+In our example setup, we wish to call our remote function ``func()`` five 
+times, and store the result of each call into a list. To compare the 
+performance of different ways of looping our calls to our remote function, 
+we can define each loop version as a separate function on the driver script.
+
+For the first version **ex1**, each iteration of the loop calls the remote 
+function, then calls ``ray.get`` in an attempt to store the current result 
+into the list, as follows:
+
+.. code-block:: python
+
+  # This loop is suboptimal in Ray, and should only be used for the sake of this example
+  def ex1():  
+    list1 = []
+    for i in range(5):
+      list1.append(ray.get(func.remote()))
+
+For the second version **ex2**, each iteration of the loop calls the remote 
+function, and stores it into the list **without** calling ``ray.get`` each time. 
+``ray.get`` is used after the loop has finished, in preparation for processing 
+``func()``'s results:
+
+.. code-block:: python
+
+  # This loop is more proper in Ray
+  def ex2():
+    list2 = []
+    for i in range(5):
+      list2.append(func.remote())
+    ray.get(list2)
+
+Finally, as a demonstration of Ray's parallelism abilities, let's create a 
```

Comment:
Oh, this example should be a bit different. The idea is to call another function (which is not remote) as other_func; this is going to slow down the loop since that part is not parallelized (so just remove the @ray.remote from other_func and also the .remote). This shows that if there is still a large serial component to the loop, parallelism speedup might not be as large as expected (this is called https://en.wikipedia.org/wiki/Amdahl%27s_law). Does it make sense? Please modify the description accordingly.